mlir/tools/miopen-gen/miopen-gen.cpp

```diff
@@ -2024,25 +2058,50 @@ int main(int argc, char **argv) {
     mod = miopenModule;
   }
 
+  assert(mod == module);
+
+
+  // Compute set of call-graph root nodes;  they're the ones we need to
+  // call from main().  Start with all nodes, then erase the ones that
+  // have edges to them.
+  CallGraph cg(mod);
+  llvm::DenseSet<CallGraphNode*> roots(cg.begin(), cg.end());
+  for (auto &node : roots)
+    for (auto &edge : *node)
+      roots.erase(edge.getTarget());
+
+  // Make KernelIFs for the roots, to pass to populateHostHarnessLogic().
+  SmallVector<KernelIF, 8> rootIFs;
+  for (auto node : roots) {
+    FuncOp func = dyn_cast<FuncOp>(node->getCallableRegion()->getParentOp());
+    rootIFs.emplace_back(func);
+  }
+
+  // Sort by name to be deterministic (because DenseSet above).
+  llvm::sort(rootIFs, [&](KernelIF &a, KernelIF &b) {
```

Comment:
There's a test that generates sub-kernels with names including "_0_", "_1_", and "_2_", and they pass results as arguments to the next one so they have to be called in the right order.  DenseSet, used to generate the roots, is not necessarily in the right order.  Could also sort by location, but it doesn't have a native ordering and converting to opaque pointer felt crude.

Reply:
Would putting together the actual def-use graph be too much effort?

Reply:
Or does this not matter for the actual IR but only for printing it?

Reply:
The def-use graph should be easy, given the underlying SSA-ness of llvm, and that would certainly be more generally correct.  Let me look into that.

Reply:
On further consideration, I think the right idea is simply to preserve the order of the roots.  The original problem case was a generated kernel, which made three kernels named _0, _1, and _2, so sorting by name kept the generated order.

Since the code before me called them in original order, and miopen-gen is not a general tool, preserving that order will at least not break anything new.

Reply:
Changed DenseSet to SetVector to keep the original order, and then I can remove the sort() call entirely.